### PR TITLE
Login screen

### PR DESCRIPTION
**Pull request by AbdulrahmanAli19**
LoginScreen.kt working as expected,

NOTE
the search query does not work when the password gets sent with "tags" so I have ended up deleting it to get the job done

### CODE REVIEW

File: app/src/main/java/com/iti/android/team1/ebuy/model/datasource/remotesource/RetrofitHelper.kt [hunk 1]
```diff
@@ -49,7 +49,8 @@ object RetrofitHelper : RemoteSource {
     }
 
     override suspend fun loginCustomer(customerLogin: CustomerLogin): Response<CustomerLoginAPI> {
-        return retrofitService.loginCustomer(customerLogin)
+        val query = "email:${customerLogin.email}"
```
Comment (AbdulrahmanAli19) [inline app/src/main/java/com/iti/android/team1/ebuy/model/datasource/remotesource/RetrofitHelper.kt:52]: when I add tags:${customerLogin.email} to the query it returns all the list :(

**Review by Mina-Wael — APPROVED** [review]

File: app/src/main/java/com/iti/android/team1/ebuy/ui/login_screen/view/LoginScreen.kt [hunk 2]
```diff
@@ -27,32 +33,65 @@ class LoginScreen : Fragment() {
...
+                    password = binding.edtPassword.text.toString().trim()
+                )
+                fetchUserData()
             } else if (!AuthRegex.isEmailValid(binding.edtEmail.editableText.toString())) {
-                ///TODO: email is week
                 binding.textInputLayout.error = "Email is not valid"
             } else if (!AuthRegex.isPasswordValid(binding.edtPassword.editableText.toString())) {
-                ///TODO: password not valid
-                binding.textInputLayout2.error = "Password is not valid"
+                binding.textInputLayout2.error = "Wrong password"
```
Comment (MohabElziny) [inline app/src/main/java/com/iti/android/team1/ebuy/ui/login_screen/view/LoginScreen.kt:64]: The logic shouldn't be in the view, Use The same sealed class as in Register Screen and put the logic on viewmodel

File: app/src/main/java/com/iti/android/team1/ebuy/ui/login_screen/viewmodel/LoginScreenViewModel.kt [hunk 3]
```diff
@@ -13,23 +13,46 @@ import kotlinx.coroutines.flow.MutableStateFlow
...
         MutableStateFlow(ResultState.Loading)
     val loginState get() = _loginState.asStateFlow()
 
-    fun makeLoginRequest(customerLogin: CustomerLogin) {
+    fun makeLoginRequest(email: String, password: String) {
+
+        val customerLogin = CustomerLogin(
+            email = email,
+            password = repository.encodePassword(password)
+        )
```
Comment (MohabElziny) [inline app/src/main/java/com/iti/android/team1/ebuy/ui/login_screen/viewmodel/LoginScreenViewModel.kt:28]: we put the encoding in repository not in viewmodel

File: app/src/main/java/com/iti/android/team1/ebuy/ui/login_screen/viewmodel/LoginScreenViewModel.kt [hunk 4]
```diff
@@ -13,23 +13,46 @@ import kotlinx.coroutines.flow.MutableStateFlow
...
+            is NetworkResponse.FailureResponse -> {
+                _loginState.emit(ResultState.Error(result.errorString))
+            }
+            is NetworkResponse.SuccessResponse -> {
+                if (password == repository.decodePassword(result.data.password
+                        ?: "")
+                )
+                    _loginState.emit(ResultState.Success(result.data))
+                else
+                    _loginState.emit(ResultState.Error("Invalid data"))
```
Comment (MohabElziny) [inline app/src/main/java/com/iti/android/team1/ebuy/ui/login_screen/viewmodel/LoginScreenViewModel.kt:53]: why do you check for password here?

File: app/src/main/java/com/iti/android/team1/ebuy/ui/register_screen/view/RegisterScreen.kt [hunk 5]
```diff
@@ -102,6 +102,8 @@ class RegisterScreen : Fragment() {
             binding.edtFirstName.text.toString().trim(),
             binding.edtLastName.text.toString().trim(),
             binding.edtPassword.text.toString().trim())
+        "noobmaster@lod.com"
+        "ASDASsdD##$123"
```
Comment (MohabElziny) [inline app/src/main/java/com/iti/android/team1/ebuy/ui/register_screen/view/RegisterScreen.kt:106]: what's this?

Comment (AbdulrahmanAli19) [reply app/src/main/java/com/iti/android/team1/ebuy/ui/register_screen/view/RegisterScreen.kt:106]: l left them to make it easy in you'd we sign in once more,

File: app/src/main/java/com/iti/android/team1/ebuy/model/datasource/localsource/prefs/PreferenceProvider.kt [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+private const val PREF_NAME = "E_BUY_APP"
+
+private const val USER_ID = "USER_ID_KEY"
+private const val IS_USER_SINGED_IN = "IS_USER_SINGED_IN"
+
+class PreferenceProvider(context: Context) {
+
+    private val sharedPref: SharedPreferences =
+        context.applicationContext.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE)
```
[the remark this answers is not in the document]
Comment (AbdulrahmanAli19) [reply app/src/main/java/com/iti/android/team1/ebuy/model/datasource/localsource/prefs/PreferenceProvider.kt:14]: okay

File: app/src/main/java/com/iti/android/team1/ebuy/ui/login_screen/view/LoginScreen.kt (hunk 2 above)
Comment (AbdulrahmanAli19) [reply app/src/main/java/com/iti/android/team1/ebuy/ui/login_screen/view/LoginScreen.kt:64]: okay

File: app/src/main/java/com/iti/android/team1/ebuy/ui/login_screen/viewmodel/LoginScreenViewModel.kt (hunk 3 above)
Comment (AbdulrahmanAli19) [reply app/src/main/java/com/iti/android/team1/ebuy/ui/login_screen/viewmodel/LoginScreenViewModel.kt:28]: ill update it

File: app/src/main/java/com/iti/android/team1/ebuy/ui/login_screen/viewmodel/LoginScreenViewModel.kt (hunk 4 above)
Comment (AbdulrahmanAli19) [reply app/src/main/java/com/iti/android/team1/ebuy/ui/login_screen/viewmodel/LoginScreenViewModel.kt:53]: the API was checking on the email only not on the password so i that's was the only way to make that this is the user, ill updated after we managed to filter the data

File: app/src/main/java/com/iti/android/team1/ebuy/ui/register_screen/view/RegisterScreen.kt (hunk 5 above)
Comment (AbdulrahmanAli19) [reply app/src/main/java/com/iti/android/team1/ebuy/ui/register_screen/view/RegisterScreen.kt:106]: oh forgot to delete them I thought that u mean the XML text

File: app/src/main/java/com/iti/android/team1/ebuy/ui/login_screen/viewmodel/LoginScreenViewModel.kt (hunk 4 above)
Comment (MohabElziny) [reply app/src/main/java/com/iti/android/team1/ebuy/ui/login_screen/viewmodel/LoginScreenViewModel.kt:53]: The API returns a response if the email and password are the same as in the API, thus you don't need to check for the password here if the answer is success, but you will emit an error if there is no customer in the response.

File: app/src/main/java/com/iti/android/team1/ebuy/ui/register_screen/view/RegisterScreen.kt (hunk 5 above)
Comment (MohabElziny) [reply app/src/main/java/com/iti/android/team1/ebuy/ui/register_screen/view/RegisterScreen.kt:106]: No, remove it from the code!

Comment (AbdulrahmanAli19) [reply app/src/main/java/com/iti/android/team1/ebuy/ui/register_screen/view/RegisterScreen.kt:106]: I just said that "oh forgot to delete them I thought that u mean the XML text"

**Comment by AbdulrahmanAli19** [comment]
All the necessary changes have been made,
new dialog has been added if the user entered the wrong password,
all the data that gets stored in the shared presence is encrypted
PreferenceProvider.kt is following a singleton pattern

**Review by MohabElziny — APPROVED** [review]

**Pull request merged**